### PR TITLE
gsil debugging 2

### PR DESCRIPTION
**Pull request by eeckstein**
#### Resolved bug number: ([SR-](https://bugs.swift.org/browse/SR-))

---

Before merging this pull request to apple/swift repository:
- [ ] Test pull request on Swift continuous integration.

<details>
  <summary>

Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please smoke test |
| OS X platform | @swift-ci Please smoke test OS X platform |
| Linux platform | @swift-ci Please smoke test Linux platform |

 **Validation Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please test |
| OS X platform | @swift-ci Please test OS X platform |
| Linux platform | @swift-ci Please test Linux platform |

Note: Only members of the Apple organization can trigger swift-ci.
</details>

### CODE REVIEW

**Comment by eeckstein** [comment]
@swift-ci Please smoke test

**Comment by eeckstein** [comment]
@swift-ci Please smoke test

File: include/swift/SIL/SILLocation.h [hunk 1]
```diff
@@ -84,10 +84,15 @@ class SILLocation {
...
-  } DebugLoc;
+  struct DebugLoc {
+    unsigned Line;
+    unsigned Column;
+    const char *Filename;
+
+    DebugLoc(unsigned Line = 0, unsigned Column = 0,
+             const char *Filename = nullptr) : Line(Line), Column(Column),
+                                               Filename(Filename) { }
+  };
```
Comment (adrian-prantl) [inline include/swift/SIL/SILLocation.h]: I'm just curious, what's the use-case for the empty DebugLoc?

File: lib/IRGen/IRGenDebugInfo.cpp [hunk 2]
```diff
@@ -661,12 +646,13 @@ llvm::DISubprogram *IRGenDebugInfo::emitFunction(
...
+  // thunk helpers, where DS->Loc is an arbitrary location of whichever use
+  // was emitted first.
   if (DS && (!SILFn || (!SILFn->isBare() && !SILFn->isThunk()))) {
-    auto FL = getFullLocation(DS->Loc, SM);
-    L = FL.Loc;
-    ScopeLine = FL.LocForLinetable.Line;
+    L = DS->Loc.decodeDebugLoc(SM);
+    ScopeLine = L.Line;
+    if (!DS->Loc.isDebugInfoLoc())
+      L = SILLocation::decode(DS->Loc.getSourceLoc(), SM);
```
Comment (adrian-prantl) [inline lib/IRGen/IRGenDebugInfo.cpp]: LGTM. Thanks!

File: include/swift/SIL/SILLocation.h (hunk 1 above)
Comment (eeckstein) [inline include/swift/SIL/SILLocation.h]: Existing code in bool SILParser::parseSILLocation(SILLocation &Loc).
It's probably better to rewrite this and not allow an empty DebugLoc.

Comment (adrian-prantl) [inline include/swift/SIL/SILLocation.h]: > On Mar 18, 2016, at 3:03 PM, eeckstein <notifications@github.com> wrote:
> 
> In include/swift/SIL/SILLocation.h <https://github.com/apple/swift/pull/1735#discussion_r56728646>:
> 
> > @@ -84,10 +84,15 @@ class SILLocation {
> >      DebugInfoKind = 1 << 3 | 1 << 4
> >    };
> >  
> > -  typedef struct {
> > -    unsigned Line = 0, Column = 0;
> > -    const char *Filename = nullptr;
> > -  } DebugLoc;
> > +  struct DebugLoc {
> > +    unsigned Line;
> > +    unsigned Column;
> > +    const char *Filename;
> > +
> > +    DebugLoc(unsigned Line = 0, unsigned Column = 0,
> > +             const char *Filename = nullptr) : Line(Line), Column(Column),
> > +                                               Filename(Filename) { }
> > +  };
> Existing code in bool SILParser::parseSILLocation(SILLocation &Loc).
> It's probably better to rewrite this and not allow an empty DebugLoc.
> 
> 
Ah yes. But rewriting involves a tedious rewriting of all call sites. We don’t necessarily need to rewrite parseSILLocation() now, but it’s probably better to not add this constructor so we don’t encourage creation of empty locations.

-- adrian

**Pull request merged**